pkg/arvo/app/dojo.hoon

```diff
@@ -1168,6 +1168,140 @@
       *tank
     ~(duck easy-print type)
   ::
+  ::  Full tab complete for all Dojo sinks and sources is a madmans job.
+  ::  Instead, we try to parse limited but common forms we know we can
+  ::  autocomplete correctly
+  ++  he-tab-not-hoon
+    |=  [pos=@ud txt=tape]
+    ^+  +>
+    =*  res  +>
+    |^
+    =/  naked-poke=(unit term)
+      %+  rust  txt
+      (full (ifix [col (just `@`10)] ;~(pose sym (easy %$))))
+    ?^  naked-poke
+      (complete-naked-poke u.naked-poke)
+    =/  variable=(unit term)
+      %+  rust  txt
+      (full (ifix [tis (just `@`10)] ;~(pose sym (easy %$))))
+    ?^  variable
+      (complete-variable u.variable)
+    =/  gen-poke-to-app=(unit [term term])
+      %+  rust  txt
+      ;~  sfix
+        ;~  (glue bar)
+          ;~(pose ;~(pfix col sym) (easy %$))
+          ;~(pose sym (easy %$))
+        ==
+        (just `@`10)
+      ==
+    ?^  gen-poke-to-app
+      (complete-gen-poke-to-app u.gen-poke-to-app)
+    =/  naked-gen=(unit term)
+      %+  rust  txt
+      (full (ifix [lus (just `@`10)] ;~(pose sym (easy %$))))
+    ?~  naked-gen
+      res
+    (complete-naked-gen u.naked-gen)
+    ::
+    ++  complete-naked-poke
+      |=  app=term
+      =/  pax=path
+        /(scot %p our.hid)/[q.byk.hid]/(scot %da now.hid)/app
+      %+  complete  (cat 3 ':' app)
+      %+  murn  ~(tap by dir:.^(arch %cy pax))
+      |=  [=term ~]
+      ^-  (unit [^term tank])
+      ?.  =(app (end 3 (met 3 app) term))
+        ~
+      ?~  =<(fil .^(arch %cy (weld pax ~[term %hoon])))
+        ~
+      `[(cat 3 ':' term) *tank]
+    ::
+    ++  complete-variable
+      |=  variable=term
+      %+  complete  variable
+      %+  murn  ~(tap by var)
+      |=  [name=term =cage]
+      ^-  (unit [term tank])
+      ?.  =(variable (end 3 (met 3 variable) name))
+        ~
+      `[name (xsell q.cage)]
+    ::
+    ++  complete-gen-poke-to-app
+      |=  [app=term gen=term]
+      =.  app
+        ?:(?=(%$ app) %hood app)
+      %+  complete
+        ?:  =(%hood app)
+          (cat 3 '|' gen)
+        :((cury cat 3) ':' app '|' gen)
+      =/  pfix=path
+        /(scot %p our.hid)/[q.byk.hid]/(scot %da now.hid)/gen/[app]
+      ::
+      %^  tab-generators:auto  pfix  `app
+      %+  murn
+        ~(tap by dir:.^(arch %cy pfix))
+      |=  [=term ~]
+      ?.  =(gen (end 3 (met 3 gen) term))
+        ~
+      ?~  =<(fil .^(arch %cy (weld pfix ~[term %hoon])))
```

Comment:
Why the inconsistency between this and the `dir` access above (where you used infix `:`)?

Reply:
The parser didn't like `:.^(` inside tall form, but was fine inside wide form.

Reply:
Huh, okay.  It would probably work if you wrap the right side with `[]`, but that's not really prettier, so this is fine.